imap_processing/hi/l1a/science_direct_event.py

```diff
@@ -163,55 +175,87 @@ def break_into_bits_size(binary_data: str):
+    packet_met_time : list
+        List of packet MET time
 
     Returns
     -------
     dataset: xarray.Dataset
         xarray dataset
     """
     # These are the variables that we will store in the dataset
-    met_subseconds = []
-    met_seconds = []
+    de_met_time = []
     esa_step = []
     trigger_id = []
     tof_1 = []
     tof_2 = []
     tof_3 = []
     de_tag = []
-    for event in de_data_list:
-        # TODO: how to check for fist event?
-        # metaevent.
-        # How to handle if first event is not
-        # metaevent? This means that current
-        # data file started with direct event.
-        # metaevent is a way to store information
-        # about bigger portion of time information. Eg.
-        # metaevent stores information about, let's say
-        # "20240319T09:30:01". Then direct event time
-        # tag stores information of time ticks since
-        # that time. Then we use those two to combine and
-        # get exact time information of each event.
+    ccsds_met = []
+
+    # How to handle if first event is not metaevent? This
+    # means that current data file started with direct event.
+    # Per Paul, log a warning and discard all direct events
+    # until we see next metaevent because it could mean
+    # that the instrument was turned off during repoint.
+
+    # Find the index of the first occurrence of the metaevent
+    first_metaevent_index = next(
+        (i for i, d in enumerate(de_data_list) if d.get("start_bitmask_data") == 0),
+        None,
+    )
+
+    if first_metaevent_index is None:
+        return None
+    elif first_metaevent_index != 0:
+        # Discard all direct events until we see next metaevent
+        # TODO: log a warning
+        de_data_list = de_data_list[first_metaevent_index:]
+        packet_met_time = packet_met_time[first_metaevent_index:]
+
+    for index, event in enumerate(de_data_list):
         if event["start_bitmask_data"] == 0:
+            # metaevent is a way to store information
+            # about bigger portion of time information. Eg.
+            # metaevent stores information about, let's say
+            # "20240319T09:30:01.000". Then direct event time
+            # tag stores information of time ticks since
+            # that time. Then we use those two to combine and
+            # get exact time information of each event.
+
             # set time and esa step values to
             # be used for direct event followed by
             # this metaevent
             int_subseconds = event["subseconds"]
             int_seconds = event["seconds"]
             current_esa_step = event["esa_step"]
+            # Add half a tick once per algorithm document
+            # and Paul Janzen.
+            half_tick = LOOKED_UP_DURATION_OF_TICK // 2
+            # convert microseconds to millieseconds to
+            # match subseconds time format
+            half_tick_ms = half_tick // 1e3
+            int_subseconds += half_tick_ms
```

Comment:
Could you add a link to the page in the algorithm document for this? Might be easier to track it down in the future.

Do you know what values you expect this to be, it doesn't seem quite right to be dividing `3999` by // 2 into an integer `1999`, do you want that to be `1999.5` instead? Then the next line produces a float half_tick_ms because of 1e3, so you might want to switch that over to 1000 instead. But also note that it is producing 1 ms, from an initial value of 3999 which is smaller than half a tick.

```
>>> 3999 // 2 // 1000
1
>>> 3999 // 2 // 1e3
1.0
```

Reply:
I will add the section from algorithm document. In the document it said like in the code block and Paul suggested it doing that. I will think about this and double check with him.
```
Take the MET(in code, it's direct event MET time), add half a tick
```
The reason he wants to add half a tick is because event could be in between ticks and he wanted to add this to account for that.

Reply:
Currently, it is closer to 1/4 of a tick if it is returning 1, so it'd be worth checking with him.

Another option is to change to `int_ns` here so you go to nanoseconds instead of milliseconds, or make milliseconds a float value instead of integer. Because in the following routine you turn the time into nanoseconds since MET.

Reply:
After looking through it closely. I made a big mistake with calculation. Looks like I was converting it ns even though my comment said I was converting it millisecond. Thank you for catching it. Will push change soon and if you could double check, I appreciate it.